### PR TITLE
Added gaussian label smoother for number tokens 

### PR DESCRIPTION
**Pull request by ad045**
### Description:
This pull request introduces a label smoothing technique that applies Gaussian smoothing exclusively to number tokens. The key changes are outlined below: 

- **Gaussian Label Smoother**: 
  - Implemented the `GaussianLabelSmoother` class in `src/ntl/utils/label_smoother.py`.
  - Controlled via two new `ModelArguments`: `gaussian_label_smoother` and `label_smoother_sigma`.

- **Number Token Selector**:
  - Extracted the token selection logic from `NumberTokenLoss` into a new `NumberTokenSelector` class located in `src/ntl/utils/number_token_selector.py`.
  - Utilized by both `GaussianLabelSmoother` and `NumberTokenLoss` to consistently select number tokens.

- **Other Changes**:
    -  Updated `src/ntl/run_language_modeling.py` and `src/ntl/trainer.py` to support the new label smoothing feature.
    - Modified `src/ntl/loss_functions/number_token_loss.py` to integrate with the new `NumberTokenSelector`.
    - Updated `src/ntl/args.py` to include new arguments for the Gaussian Label Smoother.
    - Added unit tests to ensure the correctness of the `GaussianLabelSmoother` and `NumberTokenSelector`.

### Types of Changes
- [x] **New feature** (non-breaking change which adds functionality)
- [x] **Refactor** (non-breaking change which improves the code structure)

### CODE REVIEW

File: src/ntl/utils/number_token_selector.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        self.nvocab = torch.full((vocab_size,), float("nan"), device=device)
+
+        hashed_num_tokens = set(self.tokenizer.get_num_tokens())
+        
+        for token, id in self.tokenizer.get_vocab().items():
+            if token in hashed_num_tokens:
+                self.nvocab[id] = self.tokenizer.decode_number_token(token, ignore_order=True)
+
+
+    def select_number_tokens(self, logits: Tensor, labels: Tensor):
```
Comment (jannisborn) [inline src/ntl/utils/number_token_selector.py:22]: Labels are not used in here so I would remove them from the signature

File: src/ntl/utils/label_smoother.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            logits = logits[..., :-1, :].contiguous()
+            labels = labels[..., 1:].contiguous()
+
+        # Select only number tokens for smoothing
+        if self.selector is not None:
+            if not hasattr(self.selector, 'nvocab'):
+                raise AttributeError("The selector must have an attribute 'nvocab' representing the number of valid vocab tokens.")
+    
+            # Select number tokens
+            logits, labels, number_tokens = self.selector.select_number_tokens(logits, labels)
```
Comment (jannisborn) [inline src/ntl/utils/label_smoother.py:55]: remove labels here

File: src/ntl/loss_functions/number_token_loss.py [hunk 3]
```diff
@@ -3,38 +3,30 @@
...
             raise ValueError("Logits passed to the NumberTokenLoss are empty!")
         if labels.numel() == 0:
             raise ValueError("Labels passed to the NumberTokenLoss are empty!")
 
-        labels = labels.masked_fill(labels == -100, 0)
-
-        # Create a mask to filter out non-digit tokens
-        number_tokens = ~torch.isnan(self.nvocab)
-        logits = logits[:, :, number_tokens]
+        logits, labels, number_tokens = self.selector.select_number_tokens(logits, labels)
```
Comment (jannisborn) [inline src/ntl/loss_functions/number_token_loss.py:24]: remove labels here also

**Review by jannisborn — APPROVED** [review]
Great work @ad045 ! Seems very flexible and easy to use. Super solid job on the test suite 👍🏼 

I left some cosmetics comments where unnecessary things can be remove, but conceptually we are ready to merge already now!

File: src/ntl/utils/number_token_selector.py (hunk 1 above)
Comment (jannisborn) [reply src/ntl/utils/number_token_selector.py:22]: and remove them from the return statement below then obviously
